framework/Optimizers/SimulatedAnnealing.py

```diff
@@ -0,0 +1,718 @@
+      @ In, point, dict, point to submit
+      @ In, traj, int, trajectory identifier
+      @ In, step, int, iteration number identifier
+      @ In, moreInfo, dict, optional, additional run-identifying information to track
+      @ Out, None
+    """
+    info = {}
+    if moreInfo is not None:
+      info.update(moreInfo)
+    info.update({'traj': traj,
+                  'step': step
+                })
+    # NOTE: explicit constraints have been checked before this!
+    self.raiseADebug('Adding run to queue: {} | {}'.format(point, info))
+    self._submissionQueue.append((point, info))
+  # END queuing Runs
+  # * * * * * * * * * * * * * * * *
+
+  ###############
+  # Run Methods #
+  ###############
+  def _useRealization(self, info, rlz):
+    """
+      Used to feedback the collected runs into actionable items within the sampler.
+      @ In, info, dict, identifying information about the realization
+      @ In, rlz, dict, realized realization
+      @ In, optVal, float, value of objective variable (corrected for min/max)
+      @ Out, None
+    """
+    traj = info['traj']
+    info['optVal'] = rlz[self._objectiveVar]
+    self.incrementIteration(traj)
+    self._resolveNewOptPoint(traj, rlz, rlz[self._objectiveVar], info)
+    if self._stepTracker[traj]['opt'] == None:
+      # revert to the last accepted point
+      rlz = self._optPointHistory[traj][-1][0]
+      info = self._optPointHistory[traj][-1][1]
+      info['step'] = self.getIteration(traj)
+      optVal = rlz[self._objectiveVar]
+    iter = int(self.getIteration(traj) +1) # Is that ok or should we always keep the traj in case I have multiple trajectories in parallel?
+    fraction = iter/self.limit
+    currentPoint = self._collectOptPoint(rlz)
+    self.T0 = self._temperature(fraction)
+    self.T = self._coolingSchedule(iter,self.T0)
+    if traj in self._activeTraj:
+      newPoint = self._nextNeighbour(rlz,fraction)
+      # check new opt point against constraints
+      try:
+        suggested, modded = self._handleExplicitConstraints(newPoint, currentPoint, 'opt')
+      except NoConstraintResolutionFound:
+        # we've tried everything, but we just can't hack it
+        self.raiseAMessage('Optimizer "{}" trajectory {} was unable to continue due to functional or boundary constraints.'
+                          .format(self.name, traj))
+        self._closeTrajectory(traj, 'converge', 'no constraint resolution', newPoint[self._objectiveVar])
+        return
+      self._submitRun(suggested, traj, self.getIteration(traj))
+
+  # * * * * * * * * * * * * * * * *
+  # Convergence Checks
+  # Note these names need to be formatted according to checkConvergence check!
+  convFormat = ' ... {name:^12s}: {conv:5s}, {got:1.2e} / {req:1.2e}'
+
+  # NOTE checkConvSamePoint has a different call than the others
+  # should this become an informational dict that can be passed to any of them?
+
+  def checkConvergence(self, traj, new, old):
+    """
```

Comment:
is this copied/pasted? is it needed?

Reply:
is it needed?

Reply:
yes.
converged, convDict = self.checkConvergence(traj, new, old)